yt_dlp/extractor/microsoftembed.py

```diff
@@ -63,3 +71,217 @@ def _real_extract(self, url):
             'subtitles': subtitles,
             'thumbnails': thumbnails,
         }
+
+
+class MicrosoftMediusBaseIE(InfoExtractor):
+    def _sub_to_dict(self, subtitile_list):
+        subtitles = {}
+        for sub in subtitile_list:
+            subtitles.setdefault(sub['tag'], []).append({k: v for k, v in sub.items() if k != 'tag'})
+        return subtitles
```

Comment:
Couple things: We could mutate the original dict, since we control callees.
Also typo fix and static method
```suggestion
    @staticmethod
    def _sub_to_dict(self, subtitle_list):
        subtitles = {}
        for sub in subtitle_list:
            subtitles.setdefault(sub.pop('tag', None) or 'unknown', []).append(sub)
        return subtitles
```
Side note @ maintainers, this pattern seems to come up A LOT now that traversal allows you to generate list of items. Mayhaps this could become a utils function?

Reply:
Something like this?

```py
def groupby(key_path, value_path):
    def inner(data):
        return {
            k: traverse_obj(v, (..., value_path))
            for k, v in itertools.groupby(data, lambda x: traverse_obj(x, key_path))
            if k is not None
        }
    return inner

```

Reply:
Something like this?

```py
def to_dict_by(key_path, value_path, default_key=None):
    def inner(data):
        ret = defaultdict(list)
        for d in data:
            k = traverse_obj(d, key_path, default=default_key)
            ret[k].append(traverse_obj(d, value_path))
        return dict(ret)
    return inner

```

You'd use it as:

```py
    traverse_obj(obj,(*common_path, to_dict_by(key_path, value_path)))
```

---

Alternatively, we could implement it into traverse_obj to be used like:
```py
traverse_obj(obj, (*common_path, {
    (..., key_path): value_path
))
```

<details>
<summary>Possible implementation</summary>

```diff
diff --git a/test/test_utils.py b/test/test_utils.py
index 09c648cf8..ac7a2d0a6 100644
--- a/test/test_utils.py
+++ b/test/test_utils.py
@@ -2191,6 +2191,20 @@ def test_traverse_obj(self):
         self.assertEqual(traverse_obj(_TEST_DATA, {0: ('dict', ...)}), {},
                          msg='remove key if branch in dict key not successful')

+        # Test tuple in dict as key
+        _TEST_DATA2 = [
+            {'key': 1,'value': '1.1'},
+            {'key': 1, 'value': '1.2'},
+            {'key': 2, 'value': '2.1'},
+            {'value': '0.1'},
+            {'key': 2},
+            {'key': 3},
+        ]
+        self.assertEqual(traverse_obj(_TEST_DATA2, {(..., 'key'): 'value'}), {1: ['1.1', '1.2'], 2: ['2.1']},
+                         msg='TODO')
+        self.assertEqual(traverse_obj(_TEST_DATA2, (..., {('key',): 'value'})), [{1: '1.1'}, {1: '1.2'}, {2: '2.1'}],
+                         msg='TODO')
+
         # Testing default parameter behavior
         _DEFAULT_DATA = {'None': None, 'int': 0, 'list': []}
         self.assertEqual(traverse_obj(_DEFAULT_DATA, 'fail'), None,
diff --git a/yt_dlp/utils/traversal.py b/yt_dlp/utils/traversal.py
index 8938f4c78..54829172c 100644
--- a/yt_dlp/utils/traversal.py
+++ b/yt_dlp/utils/traversal.py
@@ -89,6 +89,18 @@ def traverse_obj(
     else:
         type_test = lambda val: try_call(expected_type or IDENTITY, args=(val,))

+
+    def _iterate(obj):
+        if isinstance(obj, collections.abc.Mapping):
+            return True, obj.values()
+        elif is_iterable_like(obj) or isinstance(obj, xml.etree.ElementTree.Element):
+            return True, obj
+        elif isinstance(obj, re.Match):
+            return True, obj.groups()
+        elif traverse_string:
+            return False, str(obj)
+        return True, ()
+
     def apply_key(key, obj, is_last):
         branching = False
         result = None
@@ -116,18 +128,7 @@ def apply_key(key, obj, is_last):
                 apply_path(obj, branch, is_last)[0] for branch in key)

         elif key is ...:
-            branching = True
-            if isinstance(obj, collections.abc.Mapping):
-                result = obj.values()
-            elif is_iterable_like(obj) or isinstance(obj, xml.etree.ElementTree.Element):
-                result = obj
-            elif isinstance(obj, re.Match):
-                result = obj.groups()
-            elif traverse_string:
-                branching = False
-                result = str(obj)
-            else:
-                result = ()
+            branching, result = _iterate(obj)

         elif callable(key):
             branching = True
@@ -150,11 +151,33 @@ def apply_key(key, obj, is_last):
                 result = ''.join(result)

         elif isinstance(key, dict):
-            iter_obj = ((k, _traverse_obj(obj, v, False, is_last)) for k, v in key.items())
-            result = {
-                k: v if v is not None else default for k, v in iter_obj
-                if v is not None or default is not NO_DEFAULT
-            } or None
+            result = {}
+            for key_path, val_path in key.items():
+                dynamic_key = isinstance(key_path, tuple)
+                accumulate = dynamic_key and key_path[0] is ...
+                if accumulate:
+                    key_path = key_path[1:]
+
+                for item in _iterate(obj)[1] if accumulate else (obj,):
+                    if dynamic_key:
+                        k = _traverse_obj(item, key_path, False, False)
+                    else:
+                        k = key_path
+
+                    v = _traverse_obj(item, val_path, False, is_last)
+                    if v is None:
+                        if default is NO_DEFAULT:
+                            continue
+                        v = default
+
+                    if k is None and dynamic_key:
+                        continue
+                    elif accumulate:
+                        result.setdefault(k, []).append(v)
+                    else:
+                        result[k] = v
+
+            result = result or None

         elif isinstance(obj, collections.abc.Mapping):
             result = (try_call(obj.get, args=(key,)) if casesense or try_call(obj.__contains__, args=(key,)) else
```

</details>